contexts/ModalContext.tsx

```diff
@@ -26,15 +27,31 @@ const ModalContext = createContext<Value | null>(null);
 export const ModalContextProvider: FC<PropsWithChildren> = ({ children }) => {
   const [current, setCurrent] = useState<Identifier>();
   const [args, setArgs] = useState<Args<Identifier>>();
-  const open = useCallback((identifier: Identifier, _args: Args<Identifier>) => {
-    setArgs(_args);
-    setCurrent(identifier);
-  }, []);
+  const router = useRouter();
+  const open = useCallback(
+    (identifier: Identifier, _args: Args<Identifier>) => {
+      setArgs(_args);
+      setCurrent(identifier);
+      router.replace({
+        query: { ...router.query, [identifier]: null },
+      });
+    },
+    [router],
+  );
 
   const close = useCallback(() => {
     setArgs(undefined);
     setCurrent(undefined);
-  }, []);
+    if (!current) return;
+    const { query } = router;
+    const { modal, ...rest } = query;
+    router.replace({ ...rest, [current]: undefined });
+  }, [current, router]);
+
+  useEffect(() => {
+    const [modal] = Object.keys(router.query);
```

Comment:
What happen if it has other query param? Shouldn't you should check if the modal variable is some of the Identifier?